tests/unit/dense_dot_norm2.cpp

```diff
@@ -22,125 +22,201 @@
+		std::fill( left_data.begin(), left_data.end(), static_cast< T1 >( 0 ) );
+		std::fill( right_data.begin(), right_data.end(), static_cast< T1 >( 1 ) );
+		rc = rc ? rc : alp::buildVector( left, left_data.begin(), left_data.end() );
+		rc = rc ? rc : alp::buildVector( right, right_data.begin(), right_data.end() );
+#endif
 
-	// test 1, init
+		auto left_view_even = alp::get_view( left, alp::utils::range( 0, n, 2 ) );
+		rc = rc ? rc : alp::set( left_view_even, Scalar< T1 >( testval3 ) );  // needs an implementation
+		if( rc != SUCCESS ) {
+			std::cerr << "\t test 2 (sparse, non-standard semiring) "
+				  << "initialisation FAILED\n";
+			return;
+		}
+
+		// test 2, exec
+		Scalar< T1 > out( 0 );
+		rc = alp::dot( out, left, right, pattern_sum_if );
+		if( rc != SUCCESS ) {
+			std::cerr << "\t test 2 (sparse, non-standard semiring) dot FAILED\n";
+			return;
+		}
+
+		// test 2, check
+		if( testval3 * static_cast< T1 >( n ) != *out * 2  ) {
+			std::cerr << "\t test 2 (sparse, non-standard semiring), "
+				  << "unexpected output: " << *out << ", expected " << n
+				  << ".\n";
+			rc = FAILED;
+			return;
+		}
+	}
+
+	Scalar< int > alpha( 0 );
 	alp::Semiring<
-		alp::operators::add< double >, alp::operators::mul< double >,
+		alp::operators::add< int >, alp::operators::mul< int >,
 		alp::identities::zero, alp::identities::one
-	> ring;
-	// rc = alp::set( left, 1.5 ); // left = 1.5 everywhere
-	// rc = rc ? rc : alp::set( right, -1.0 );
-	// if( rc != SUCCESS ) {
-	// 	std::cerr << "\t test 1 (dense, regular semiring): initialisation FAILED\n";
-	// 	return;
-	// }
-	Scalar< double > out( 2.55 );
-
-	// test 1, exec
-	rc = alp::dot( out, left, right, ring );
-	if( rc != SUCCESS ) {
-		std::cerr << "\t test 1 (dense, regular semiring): dot FAILED\n";
-		return;
+	> intRing;
+
+	{
+		// test 3, init
+		rc = SUCCESS;
+
+		alp::Vector< int > x( n ), y( n );
+
+#ifndef TEMP_DISABLE_SET
+		rc = alp::set( x, Scalar< T1 >( 1 ) );
+		rc = rc ? rc : alp::set( y, Scalar< T1 >( 2 ) );
+#else
+		// temp initialization
+		std::vector< T1 > x_data( n ), y_data( n );
+		std::fill( x_data.begin(), x_data.end(), 1 );
+		std::fill( y_data.begin(), y_data.end(), 2 );
+		rc = rc ? rc : alp::buildVector( x, x_data.begin(), x_data.end() );
+		rc = rc ? rc : alp::buildVector( y, y_data.begin(), y_data.end() );
+#endif
+
+		if( rc != alp::SUCCESS ) {
+			std::cerr << "\t test 3 (dense integer vectors) initialisation FAILED\n";
+			return;
+		}
+
+		// test 3, exec
+		rc = alp::dot( alpha, x, y, intRing );
+		if( rc != alp::SUCCESS ) {
+			std::cerr << "\t test 3 (dense integer vectors) dot FAILED\n";
+			return;
+		}
+
+		// test 3, check
+		if( *alpha != 2 * static_cast< int >( n ) ) {
+			std::cerr << "\t test 3 (dense integer vectors) unexpected value "
+				  << *alpha << ", expected 2 * n = " << ( 2 * n) << ".\n";
+			rc = FAILED;
+			return;
+		}
+	}
+
+	{
+		// test 4, init
+		alp::Vector< int > empty_left( 0 ), empty_right( 0 );
```

Comment:
I think it is more readable if we declare each variable in a separate line, like:
```
alp::Vector< int > empty_left( 0 );
alp::Vector< int > empty_right( 0 );
```